src/plan.rs

```diff
@@ -0,0 +1,1251 @@
+    }
+
+    fn lookup_raw_pkh_pk<Ctx: ScriptContext>(&self, hash: &hash160::Hash) -> Option<usize> {
+        Satisfier::lookup_raw_pkh_pk(self, hash).map(|p| Ctx::pk_len(&p))
+    }
+
+    fn lookup_raw_pkh_x_only_pk(&self, hash: &hash160::Hash) -> bool {
+        Satisfier::lookup_raw_pkh_x_only_pk(self, hash).is_some()
+    }
+
+    fn lookup_raw_pkh_ecdsa_sig<Ctx: ScriptContext>(&self, hash: &hash160::Hash) -> Option<usize> {
+        Satisfier::lookup_raw_pkh_ecdsa_sig(self, hash).map(|(p, _)| Ctx::pk_len(&p))
+    }
+
+    fn lookup_raw_pkh_tap_leaf_script_sig(
+        &self,
+        hash: &(hash160::Hash, TapLeafHash),
+    ) -> Option<usize> {
+        Satisfier::lookup_raw_pkh_tap_leaf_script_sig(self, hash).map(|(_, s)| s.to_vec().len())
+    }
+
+    fn lookup_sha256(&self, hash: &Pk::Sha256) -> bool {
+        Satisfier::lookup_sha256(self, hash).is_some()
+    }
+
+    fn lookup_hash256(&self, hash: &Pk::Hash256) -> bool {
+        Satisfier::lookup_hash256(self, hash).is_some()
+    }
+
+    fn lookup_ripemd160(&self, hash: &Pk::Ripemd160) -> bool {
+        Satisfier::lookup_ripemd160(self, hash).is_some()
+    }
+
+    fn lookup_hash160(&self, hash: &Pk::Hash160) -> bool {
+        Satisfier::lookup_hash160(self, hash).is_some()
+    }
+
+    fn check_older(&self, s: Sequence) -> bool {
+        Satisfier::check_older(self, s)
+    }
+
+    fn check_after(&self, l: LockTime) -> bool {
+        Satisfier::check_after(self, l)
+    }
+}
+
+/// Enum defining the type of signature required
+#[derive(Debug, Clone)]
+pub enum RequiredSig<'pk, Pk: MiniscriptKey> {
+    /// ECDSA (legacy or Segwit-v0) signature
+    Ecdsa(&'pk Pk),
+    /// Schnorr key-spend signature (BIP-341)
+    SchnorrTapKey(&'pk Pk),
+    /// Schnorr script-spend signature (BIP-341)
+    SchnorrTapScript(&'pk Pk, &'pk TapLeafHash),
+}
+
+/// Enum defining the type of preimage required
+#[derive(Debug, Clone)]
+pub enum RequiredPreimage<'h, Pk: MiniscriptKey> {
+    /// HASH160 preimage
+    Hash160(&'h <Pk as MiniscriptKey>::Hash160),
+    /// RIPEMD160 preimage
+    Ripemd160(&'h <Pk as MiniscriptKey>::Ripemd160),
+    /// HASH256 preimage
+    Hash256(&'h <Pk as MiniscriptKey>::Hash256),
+    /// SHA256 preimage
+    Sha256(&'h <Pk as MiniscriptKey>::Sha256),
+}
+
+/// Representation of a particular spending path on a descriptor. Contains the witness template
+/// and the timelocks needed for satisfying the plan.
+/// Calling `get_plan` on a Descriptor will return this structure,
+/// containing the cheapest spending path possible (considering the `Assets` given)
+#[derive(Debug, Clone)]
+pub struct Plan<'d> {
+    /// This plan's witness template
+    pub template: Vec<Placeholder<DefiniteDescriptorKey>>,
+    /// The absolute timelock this plan uses
+    pub absolute_timelock: Option<LockTime>,
+    /// The relative timelock this plan uses
+    pub relative_timelock: Option<Sequence>,
+
+    pub(crate) descriptor: &'d Descriptor<DefiniteDescriptorKey>,
```

Comment:
It seemed wasteful to clone but I think you are right, I didn't consider what would happen in multi-threaded scenarios.

Maybe I can still get away with just copying what we need, I initially started implementing that but then I had to keep updating the struct because as I worked I realized I needed more and more stuff. So I decided to just include the descriptor initially, now we can asses what we actually need (assuming we don't want to make large architectural changes) and potentially remove it.